### ISSUE TITLE
output from neuron traced model significantly different to baseline pytorch

### ISSUE THREAD

**Issue by mccowani** [aws/aws-neuron-sdk#244]
To port a custom embedding network to run on neuron, I have followed the pytorch neuron [getting started example tutorial here](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/neuron-frameworks/pytorch-neuron/tutorials/getting_started.html#pytorch-getting-started).

The model is analysed and traces, with all modules reported as supported in the debug log from neuron-cc.  I then run inference to create a set of embedding vectors.  I have followed the tutorial using an inf1.2xlarge with Deep Learning AMI and confirmed that neuron rt is running (i.e. inference is not occurring on the CPU).  The inference completes successfully with time reduction and outputs the expected vector dimensions.

However when I test the output embedding vectors for a given input compared to the baseline pytorch model output, **they differ significantly - numpy assert almost equal fails even at 1 decimal place.  See e.g. assert output below.**

The result of this in the end application is a large reduction in performance of 10-20%, making it unusable.

The model is mainly composed of convolutions and linear layers, with adaptive_avg_pool2d being the only out of the ordinary layer. The final layer in the embedding network is a Leaky ReLU layer.  

I realise neuron may cause some resolution loss, however I would expect the output of the baseline and traced models to be equal to at least several decimal places.

I have updated all tools, read all the docs, tried a range of options to the trace command including passing in a list for check_inputs, specifying check_tolerance, different optlevels, etc.  I have tried passing different example inputs to the trace, including different real data examples as well as just all zeros as per the tutorial.

Can you offer any advice to help me troubleshoot or resolve this?

torch-neuron 1.7.1.1.1.7.0
Neuron Compiler version 1.1.7.0+6d41df113

`Arrays are not almost equal to 3 decimals

Mismatched elements: 7943589 / 18433024 (43.1%)
Max absolute difference: 0.53520757
Max relative difference: 65056194.21228568
 x: array([[-2.579e-03, -1.854e-03, -2.121e-03, ..., -1.137e-03, -1.190e-03,
         1.738e-01],
       [-2.579e-03, -7.820e-04, -2.121e-03, ..., -1.030e-03, -2.728e-04,...
 y: array([[-2.581e-03, -2.070e-03, -2.117e-03, ..., -1.122e-03, -1.237e-03,
         2.000e-01],
       [-2.581e-03, -1.082e-03, -2.117e-03, ..., -1.176e-03, -4.736e-04,...`
 
@jdupl123

**Comment by aws-diamant** [comment]
Hi @mccowani 

Thanks for reporting this issue.
The diff that you're reporting is indeed unexpected and too high, we typically see relative difference of ~1% at the tensor level, which maps to close-to-zero accuracy impact at the application level.

A good starting point to debug this is to use our --fp32-cast=matmult compiler flag, which is used to selectively cast only inputs to MatMul/Conv operators to reduced precision, while performing all other ops at fp32. More documentation on this can be found [here](https://github.com/aws/aws-neuron-sdk/blob/master/docs/technotes/performance-tuning.md#mixed-precision).

If this doesn't work, would it be possible for you to share your model-setup with me, so that I can recreate the issue and debug it with you?
(feel free to share directly to aws-neuron-support@amazon.com, if that's easier than posting here).

If sharing is not a possibility, we can work together to debug this remotely, likely through starting with a smaller model and gradually adding the operators back one by one.

Looking forward to working with you on this,
Ron

**Comment by mccowani** [comment]
Thanks for the response.   I did already try the --fp32-cast=matmult compiler flag before the original post.  This gave output that was a bit closer to the original but still significantly mismatched (36% mismatched to 3 dp instead of 43%).

I will reach out directly to share further details.

**Comment by aws-joshim** [comment]
We ran multiple accuracy tests with the model sample code, using different inputs to recreate the failures. Our findings indicate that the accuracy mismatch is due to the wide dynamic range of values in the tensors. The best accuracy results were achieved with –fp32-cast=matmult-bf16 in our test.
We also found that the model test passes by normalizing the final outputs (e.g. via a softmax layer). If softmax is not workable for the model use case, other alternatives to leaky_relu depending on the purpose of your model would be to use tanh or sigmoid activations.

**Comment by jeffhataws** [comment]
Hi @mccowani, would using --fp32-cast=matmult-bf16 and normalizing the final outputs satisfy your requirements? Please let us know if we can be of further help.

**Comment by jdupl123** [comment]
Hi @jeffhataws @mccowani is currently on leave. We have done some further testing but I'll let him update you when he gets back.

**Comment by mccowani** [comment]
Hi @jeffhataws @aws-joshim Thanks for your support on this issue - apologies for the slow response.  I have validated in our internal benchmark task that the model performance now aligns with the non-neuron model so we are happy for you to consider this closed.

**Comment by jeffhataws** [comment]
It's great to hear from you and thanks for the good news. Closing ticket.
